### PR TITLE
Cac refactor

### PR DESCRIPTION
**Pull request by r0qs**
This PR refactor the chunk construction methods and remove some duplicated code.

### CODE REVIEW

File: pkg/cac/cac.go [hunk 1]
```diff
@@ -2,25 +2,59 @@ package cac
...
+var (
+	errTooShortChunkData = errors.New("short chunk data")
+	errTooLargeChunkData = errors.New("data too large")
+)
+
+// New creates a new content address chunk by initializing a span and appending the data to it.
 func New(data []byte) (swarm.Chunk, error) {
-	hasher := bmtpool.Get()
-	defer bmtpool.Put(hasher)
+	span := make([]byte, swarm.SpanSize)
```
Comment (acud) [inline pkg/cac/cac.go:18]: you should do the input validation the same way you do in `NewWithDataSpan`, since `data` can be larger than a single chunk, which should result in an error

File: pkg/cac/cac.go [hunk 2]
```diff
@@ -2,25 +2,59 @@ package cac
...
+}
+
+// NewWithSpan creates a new chunk prepending the given span to the data.
+func NewWithSpan(data []byte, span []byte) (swarm.Chunk, error) {
+	h := hasher(data)
+	hash, err := h(span)
 	if err != nil {
 		return nil, err
 	}
+	return swarm.NewChunk(swarm.NewAddress(hash), append(span, data...)), nil
```
Comment (acud) [inline pkg/cac/cac.go:43]: This is a bit dangerous. Since `span` can actually be a slice which references storage that is not guaranteed to be immutable, you can potentially end up with a data race here, this is why the race detector complains. The problem is that you're reslicing data when calling this function from line 33. The slice you're shuffling around can actually cause all sorts of quirks. I would copy at least the span in `NewWithDataSpan` to a new slice, so that when you append `data` you actually append it to a new slice, preventing all sorts of peculiar bugs in the future.

**Review by acud — CHANGES_REQUESTED** [review]

File: pkg/cac/cac.go [hunk 3]
```diff
@@ -2,25 +2,70 @@ package cac
...
+var (
+	errTooShortChunkData = errors.New("short chunk data")
+	errTooLargeChunkData = errors.New("data too large")
+)
+
+// New creates a new content address chunk by initializing a span and appending the data to it.
 func New(data []byte) (swarm.Chunk, error) {
-	hasher := bmtpool.Get()
-	defer bmtpool.Put(hasher)
+	if len(data) > swarm.ChunkSize+swarm.SpanSize {
```
Comment (acud) [inline pkg/cac/cac.go:18]: actually, here `data` can only be `0<len<=swarm.ChunkSize`, since the span is created separately below. A unit test that checks these edge cases would be good.

File: pkg/cac/cac_test.go [hunk 4]
```diff
@@ -21,3 +23,68 @@ func TestCac(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !c.Address().Equal(address) {
+		t.Fatalf("address mismatch. got %s want %s", c.Address().String(), address.String())
+	}
+}
+
+var testChunkData = []struct {
```
Comment (acud) [inline pkg/cac/cac_test.go:60]: You can inline this into the actual test below. See [here](https://github.com/ethersphere/bee/blob/master/pkg/encryption/mock/mock_test.go#L18)

File: pkg/cac/cac_test.go [hunk 5]
```diff
@@ -21,3 +23,68 @@ func TestCac(t *testing.T) {
 		t.Fatalf("address mismatch. got %s want %s", c.Address().String(), address.String())
 	}
 }
+
+func TestNewWithDataSpan(t *testing.T) {
+	foo := "greaterthanspan"
```
Comment (acud) [inline pkg/cac/cac_test.go:28]: I think that having table-drive tests for all constructors would be good, so that we can validate all input and outputs.
Note that the original sanity test that I wrote here checks only for the address of the generated chunk, however it should also check the actual `Data()` itself is correct (check that it equals `span|data`). There's no need to do this on all tests, I think that on the sanity check would suffice.

File: pkg/cac/export_test.go [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
```
Comment (acud) [inline pkg/cac/export_test.go:1]: 2021 :)

File: pkg/feeds/putter.go [hunk 7]
```diff
@@ -57,21 +57,7 @@ func (u *Putter) Put(ctx context.Context, i Index, at int64, payload []byte) err
 }
 
 func toChunk(at uint64, payload []byte) (swarm.Chunk, error) {
-	hasher := bmtpool.Get()
-	defer bmtpool.Put(hasher)
-
-	ts := make([]byte, 8)
+	ts := make([]byte, swarm.SpanSize)
```
Comment (acud) [inline pkg/feeds/putter.go:60]: note that `ts` is a timestamp here, so it is indeed `8` bytes because of the fact it is a Uint64, but I would prefer to initialize the variable with a hardcoded `8` bytes since using the `const` in `swarm` package might mislead the reader to think this is somehow related to the chunk span.

File: pkg/soc/soc.go [hunk 8]
```diff
@@ -154,7 +154,7 @@ func FromChunk(sch swarm.Chunk) (*Soc, error) {
 	spanBytes := chunkData[cursor : cursor+swarm.SpanSize]
 	cursor += swarm.SpanSize
 
-	ch, err := contentAddressedChunk(chunkData[cursor:], spanBytes)
+	ch, err := cac.NewWithSpan(chunkData[cursor:], spanBytes)
```
Comment (acud) [inline pkg/soc/soc.go:157]: You could potentially simplify this with removing `spanBytes` and the cursor increment on `L154-155`, then calling `cac.NewWithDataSpan(chunkData[cursor:])`. This way you can unexport `NewWithSpan` since it is only used here (and simplify the tests as a result).

**Review by acud — CHANGES_REQUESTED** [review]
Few more small changes. Thanks for helping us maintaining this important part of the codebase :)

File: pkg/cac/cac.go (hunk 3 above)
Comment (r0qs) [reply pkg/cac/cac.go]: I fixed it now and it is already tested in the new `TestChunkInvariants`. But currently I'm only aware of two cases:
1. Data is smaller than the spanSize
2. Data is bigger than the maxChunkSize

Do you see more cases?

File: pkg/soc/soc.go (hunk 8 above)
Comment (r0qs) [reply pkg/soc/soc.go]: True! Thanks. I fixed it now.

File: pkg/cac/cac_test.go (hunk 5 above)
Comment (r0qs) [reply pkg/cac/cac_test.go]: I could also change the sanity tests to a table-driven test and reduce some repetition in those tests, but I'm not sure if this will be good in terms of legibility.

File: pkg/cac/cac_test.go [hunk 9]
```diff
@@ -1,23 +1,101 @@
...
+			{
+				name:    "too large data chunk",
+				data:    []byte(strings.Repeat("a", swarm.ChunkSize+swarm.SpanSize+1)),
+				wantErr: cac.ErrTooLargeChunkData,
+			},
+		} {
+			testName := fmt.Sprintf("%s-%s", f.name, cc.name)
+			t.Run(testName, func(t *testing.T) {
+				_, err := f.chunker(cc.data)
+				if err != nil {
```
Comment (acud) [inline pkg/cac/cac_test.go:91]: Since you always expect errors here, I would:
```go
_,err := f.chunker()
if !errors.Is(err,cc.wantErr) { .... } 
```
Then you don't need to handle the `else` block, since `errors.Is` will return false, since `err` would be `nil`

File: pkg/cac/cac_test.go [hunk 10]
```diff
@@ -1,23 +1,101 @@
...
+		},
+	}
+
+	for _, f := range chunkerFunc {
+		for _, cc := range []struct {
+			name    string
+			data    []byte
+			wantErr error
+		}{
+			{
```
Comment (acud) [inline pkg/cac/cac_test.go:77]: I would add test cases for:
- data is a `0` length slice
- data is `nil`

File: pkg/soc/soc_test.go [hunk 11]
```diff
@@ -25,8 +26,8 @@ func TestToChunk(t *testing.T) {
 
 	id := make([]byte, 32)
 
-	payload := []byte("foo")
-	ch, err := chunk(payload)
+	payload := []byte("greaterthanspan")
```
Comment (acud) [inline pkg/soc/soc_test.go:29]: This actually does not need to be longer than span, since you're calling `New`

File: pkg/soc/soc_test.go [hunk 12]
```diff
@@ -92,8 +93,8 @@ func TestFromChunk(t *testing.T) {
...
```
Comment (acud) [inline pkg/soc/soc_test.go:96]: here too

File: pkg/soc/soc_test.go (hunk 11 above)
Comment (r0qs) [reply pkg/soc/soc_test.go]: Oh, I just noticed that I wrongly assumed that in the `New` the data also should be smaller than the span. Reading carefully your comments again I noticed that this is only the case for the `NewWithDataSpan`. I adjusted the code and the tests for both cases now.

**Review by acud — APPROVED** [review]
Brilliant! :clap: Thank you very much for this contribution :pray:

**Review by janos — APPROVED** [review]
Very nice. 👏
